File: app/routers/save.py
"""Save/load API router — event history, save/load game state."""
from fastapi import APIRouter, HTTPException

from app.database import get_db
from app.repositories import game_repo
from app.models.save import SaveListResponse, SaveSlotInfo, SaveLoadRequest, SaveLoadResponse
from app.models.player import PlayerState
from app.services.game_service import list_saves, load_save, delete_save

router = APIRouter(prefix="/api/v1/game", tags=["save"])


@router.get("/events/{session_id}")
async def get_event_history(session_id: str):
    """获取事件历史"""
    conn = get_db()
    try:
        player = game_repo.load_player(conn, session_id)
        if player is None:
            raise HTTPException(status_code=404, detail="Game session not found")
        events = game_repo.get_event_logs(conn, session_id)
        return events
    finally:
        conn.close()


@router.get("/saves", response_model=SaveListResponse)
async def get_saves(user_id: str):
    saves = list_saves(user_id)
    return SaveListResponse(
        saves=[SaveSlotInfo(**s) for s in saves]
    )


@router.post("/save/load", response_model=SaveLoadResponse)
async def load_save_endpoint(request: SaveLoadRequest):
    try:
        state = load_save(request.user_id, request.save_slot)
    except ValueError:
        raise HTTPException(status_code=404, detail="Save not found")

    player_state = PlayerState(
        id=state["session_id"],
        name=state.get("name", ""),
        gender=state.get("gender", ""),
        talent_ids=state.get("talent_ids", []),
        root_bone=state["attributes"].get("rootBone", 0),
        comprehension=state["attributes"].get("comprehension", 0),
        mindset=state["attributes"].get("mindset", 0),
        luck=state["attributes"].get("luck", 0),
        realm=state.get("realm", ""),
        realm_progress=state.get("realm_progress", 0.0),
        cultivation=state.get("cultivation", 0.0),
        age=state.get("age", 0),
        health=state.get("health", 100.0),
        qi=state.get("qi", 0.0),
        lifespan=state.get("lifespan", 100),
        faction=state.get("faction", ""),
        spirit_stones=state.get("spirit_stones", 0),
        techniques=state.get("techniques", []),
        inventory=state.get("inventory", []),
        event_count=state.get("event_count", 0),
        score=state.get("score", 0),
        technique_grades=state.get("technique_grades", []),
        ascended=state.get("ascended", False),
        is_alive=state.get("is_alive", True),
        user_id=state.get("user_id"),
        save_slot=state.get("save_slot", 0),
    )
    return SaveLoadResponse(session_id=state["session_id"], state=player_state)


@router.delete("/save/{user_id}/{save_slot}")
async def delete_save_endpoint(user_id: str, save_slot: int):
    delete_save(user_id, save_slot)
    return {"success": True}

File: web/src/api/save.ts
import { apiRequest } from './client'
import { normalizeFromPydantic } from '@/core/normalize'
import type { NormalizedGameState, PlayerStatePydantic, SaveSlotInfo, EventHistoryEntry } from '@/core/types'

interface SaveListResponse {
  saves: Array<{
    slot: number
    session_id: string
    name: string
    realm: string
    age: number
    event_count: number
    last_active_at: string | null
    is_alive: boolean
  }>
}

interface LoadSaveResponse {
  session_id: string
  state: PlayerStatePydantic
}

interface EventHistoryRawItem {
  player_id: string
  event_index: number
  event_type: string
  narrative: string
  realm: string | null
  options: unknown  // backend deserializes JSON → may be array or string
  chosen_option_id: number | null
  consequences: unknown  // backend deserializes JSON → may be object or string
  aftermath: string | null
}

export async function listSaves(userId: string): Promise<SaveSlotInfo[]> {
  const res: SaveListResponse = await apiRequest(`/api/v1/game/saves?user_id=${encodeURIComponent(userId)}`)
  return res.saves.map((s) => ({
    slot: s.slot,
    sessionId: s.session_id,
    name: s.name,
    realm: s.realm,
    age: s.age,
    eventCount: s.event_count,
    lastActiveAt: s.last_active_at,
    isAlive: s.is_alive,
  }))
}

export async function loadSave(
  userId: string,
  saveSlot: number,
): Promise<{ sessionId: string; state: NormalizedGameState }> {
  const res: LoadSaveResponse = await apiRequest('/api/v1/game/save/load', {
    method: 'POST',
    body: JSON.stringify({ user_id: userId, save_slot: saveSlot }),
  })
  return {
    sessionId: res.session_id,
    state: normalizeFromPydantic(res.state),
  }
}

export async function deleteSave(userId: string, saveSlot: number): Promise<{ success: boolean }> {
  return await apiRequest(`/api/v1/game/save/${encodeURIComponent(userId)}/${saveSlot}`, {
    method: 'DELETE',
  })
}

export async function getEventHistory(sessionId: string): Promise<EventHistoryEntry[]> {
  const res: EventHistoryRawItem[] = await apiRequest(`/api/v1/game/events/${encodeURIComponent(sessionId)}`)
  return res.map((e) => {
    let parsedAftermath: EventHistoryEntry['aftermath'] = null
    if (e.aftermath) {
      try {
        parsedAftermath = typeof e.aftermath === 'string'
          ? JSON.parse(e.aftermath)
          : e.aftermath
      } catch {
        parsedAftermath = null
      }
    }
    return {
      eventIndex: e.event_index,
      eventType: e.event_type,
      narrative: e.narrative,
      realm: e.realm,
      options: e.options,
      chosenOptionId: e.chosen_option_id,
      consequences: e.consequences,
      aftermath: parsedAftermath,
    }
  })
}
